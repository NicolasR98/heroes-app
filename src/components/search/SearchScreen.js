import React, { useMemo } from 'react'
import { useLocation, useNavigate } from 'react-router-dom';
import queryString from 'query-string'

import { useForm } from '../../hooks/useForm';
import { getHeroesByName } from '../../selectors/getHeroesByName';
import { HeroCard } from '../heroes/HeroCard';

export const SearchScreen = () => {
    const navigate = useNavigate();
    const location = useLocation();

    const { q = '' } = queryString.parse(location.search);

    const [{ searchText }, handleInputChange] = useForm({
        searchText: q,
    });

    const handleSearch = (e) => {
        e.preventDefault();
        navigate(`?q=${searchText}`);
    };

    const heroesFiltered = useMemo(() => getHeroesByName(q), [q]);

    return (
        <>
            <h1>Searchs</h1>
            <hr />
            <div className='row'>
                <div className='col-6'>
                    <h4>Search</h4>
                    <hr />
                    <form onSubmit={handleSearch}>
                        <input
                            type='text'
                            placeholder='Search a hero'
                            className='form-control'
                            name='searchText'
                            autoComplete='off'
                            onChange={handleInputChange}
                            value={searchText}
                        />

                        <button
                            className='btn btn-outline-primary mt-2'
                        >
                            Search...
                        </button>
                    </form>
                </div>

                <div className='col-6'>
                    <h4>Results</h4>
                    <hr />
                    {
                        (q === '')
                            ? <div className='alert alert-info'>Search a hero!</div>
                            : (heroesFiltered.length === 0)
                            && <div className='alert alert-danger'>No results of: {q}</div>
                    }
                    {
                        heroesFiltered.map(hero => (
                            <HeroCard
                                key={hero.id}
                                {...hero}
                            />
                        ))
                    }
                </div>

            </div>
        </>
    );
};
