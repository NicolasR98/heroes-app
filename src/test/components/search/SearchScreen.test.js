import { mount } from "enzyme";
import { MemoryRouter } from "react-router-dom";
import { SearchScreen } from "../../../components/search/SearchScreen";

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate,
}));

describe('Tests on <SearchScreen />', () => {
    test('should render correctly', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <SearchScreen />
            </MemoryRouter>
        );

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-info').text().trim()).toBe('Search a hero!');
    });

    test('should show Batman and the input value of the queryString', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchScreen />
            </MemoryRouter>
        );

        expect(wrapper.find('input').prop('value')).toBe('batman');
        expect(wrapper.find('img').prop('alt')).toBe('Batman');
    });

    test('should display an error if hero not found', () => {
        const query = 'batman!!!'
        
        const wrapper = mount(
            <MemoryRouter initialEntries={[`/search?q=${query}`]}>
                <SearchScreen />
            </MemoryRouter>
        );

        expect(wrapper.find('.alert-danger').text().trim()).toBe(`No results of: ${query}`)
    });

    test('should call `navigate` to the new screen', () => {
       const wrapper = mount(
           <MemoryRouter initialEntries={[`/search`]}>
               <SearchScreen />
           </MemoryRouter>
       );

       const expected = 'batman'

       const event = {
           target: {
               name: 'searchText',
               value: expected,
           },
       };

       wrapper.find('input').simulate('change', event);

       wrapper.find('form').prop('onSubmit')({
           preventDefault(){},
       });

       expect(mockNavigate).toHaveBeenCalledWith(`?q=${expected}`);
    });
    
});
