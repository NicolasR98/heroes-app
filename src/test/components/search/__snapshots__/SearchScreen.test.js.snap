// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on <SearchScreen /> should render correctly 1`] = `
Array [
  <h1>
    Searchs
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-6"
    >
      <h4>
        Search
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Search a hero"
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-primary mt-2"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-6"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a hero!
      </div>
    </div>
  </div>,
]
`;
